gbn: set resend timeout dynamically

ViktorTigerstrom opened the pull request:
Based on #87

This PR updates the resend timeout to be dynamically set based on how long it took for the other party to respond during the handshake process.
The timeout is set to the time it took for the server to respond multiplied by the resendMultiplier, unless the duration is shorter than the default resend timeout. If the the resend timeout has been manually set, the resend timeout will always be set to that value, and won't be dynamically set.

Prior to this PR, the timeout before a client resends the queue of packets was always a fixed value. This fixed timeout isn't suitable for all clients as the latency for different clients varies.

ViktorTigerstrom reviewed (commented):
Leaving a review comment regarding the queue `handshakeTimeout` below.

Also in general, when making this PR, one alternative approach I considered, was to continuously update resend timeout throughout the lifespan of the connection, and not just set it during the handshake process. I.e. we could instead continuously set it based on how long it took to get a response for random packets we send with some set frequency. Do you think that approach would make more sense? The benefit of that approach would be that resend timeout will reflect if the latency changes throughout the lifetime of the connection, which might be good if we have really long lived connections.

ViktorTigerstrom commented:
Rebased on the latest version of #87l

ellemouton commented on gbn/gbn_conn.go:
```diff
@@ -19,9 +19,10 @@ var (
 
 const (
 	DefaultN                = 20
-	defaultHandshakeTimeout = 100 * time.Millisecond
```
yeah I think that was an oversight - feel free to fix :)

ViktorTigerstrom replied:
Fixed with the latest push :)

ellemouton commented on gbn/gbn_conn.go:
```diff
@@ -21,7 +21,7 @@ var (
 	DefaultN                = 20
 	defaultHandshakeTimeout = 100 * time.Millisecond
-	defaultResendTimeout    = 100 * time.Millisecond
+	defaultResendTimeout    = 2000 * time.Millisecond
```
2 seconds might be a little bit long especially if we will only ever increase this....

Why do we only allow increasing? I think there is value with decreasing too and then just having a sane minimum

ViktorTigerstrom replied:
Changed this to 1 second instead, and modified so that that we can update the resend timeout dynamically to any value, as long as it's above the default minimum value!

ellemouton reviewed (commented):
I think we can make things a bit more modular and generic here. 

I think it would be cool do having something like: 

```
type TimeoutManager struct {
    ....
}

NewTimeOutManager(defaultTimeout, useStaticTimeout, multiplier, ...) 

(m *TimeoutManager) Sent() {}
(m *TimeoutManager) Received() {}
(m *TimeoutManager) GetTimeout() time.Duration() {}

``` 

Then we can use it as follows: 
- everywhere in GBN where we want to start a new timer, query the new manager's `GetTimeout` method to get what time to use. 
- as is done in this PR, do 1 `Sent/Received` cycle during the handshake. 
- Then, for each new sequence number we send & when we get the corresponding ACK - do a Sent/Received cycle. 

This will be nice cause then the GBN doesnt need to have all this logic in it and can just trust what the timeout manager says :)

ViktorTigerstrom commented:
Great idea regarding modularising it to a Timeout manager @ellemouton! 
I've updated this PR to a new TimeoutManager struct, which handles all different timeouts throughout the gbn package.
Currently it only updates the resend timeout with the handshake messages, but can be easily updated to update it when we send/receive the other messages as well.
I hope this addresses this the way you ment in your feedback :).

ViktorTigerstrom commented:
Rebased on #87

ellemouton reviewed (commented):
Nice addition! Gonna be a game changer!

Left a few comments. Also:
- does this need to be based on #87? afaikt it doesnt depend on that change - so perhaps better to decouple them since it isnt clear which one will get in first yet. 
- A test for TimeoutMgr should please be added :)

ViktorTigerstrom commented on gbn/timeout_manager.go:
```diff
@@ -62,11 +73,76 @@ func NewTimeOutManager(isServer bool) *TimeoutManager {
+
+	multipliedTimeout := time.Duration(m.resendMultiplier) * responseTime
+
+	if multipliedTimeout > defaultResendTimeout {
```
Good idea, agree! Let me know if you'd like me to change the minimum value as well, as I set both to 1sec now.

ViktorTigerstrom commented on gbn/timeout_manager.go:
```diff
@@ -62,11 +73,76 @@ func NewTimeOutManager(isServer bool) *TimeoutManager {
+// resendMultiplier.
+// If a static timeout has been manually set, then this function will do be a
+// no-op.
+func (m *TimeoutManager) updateResendTimeout(responseTime time.Duration) {
```
It currently could, but since we're moving to updating the resend timeout based on more messages then just the SYN & SYNACK messages I think it makes sense to keep this as an arg you pass to function, as the way you calculate how long the response took will vary based on different message types.

ViktorTigerstrom commented on gbn/timeout_manager.go:
```diff
@@ -29,6 +34,12 @@ type TimeoutManager struct {
+	// resendTimeout dynamically during the handshake.
+	resendMultiplier   int
+	resendMultiplierMu sync.RWMutex
+
```
That's a good idea, thanks! I'm learning a lot by this :)! I'll work on fixing that tomorrow.

ViktorTigerstrom commented on gbn/gbn_conn.go:
```diff
@@ -366,6 +366,9 @@ func (g *GoBackNConn) sendPacket(ctx context.Context, msg Message) error {
 		return fmt.Errorf("error calling sendToStream: %s", err)
 	}
 
+	// Notify the timeout manager that a message has been sent.
+	g.timeoutManager.Sent(msg)
+
```
Ah good catch! The `g.timeoutMgr.Receive(msg)` must have been lost in my previous rebase. I've added the corresponding `g.timeoutMgr.Receive(msg)` now :)

ViktorTigerstrom commented on gbn/timeout_manager.go:
```diff
@@ -0,0 +1,142 @@
+	// recvTimeout defines the max time we will wait to receive a msg before
+	// timing out.
+	recvTimeout   time.Duration
+	recvTimeoutMu sync.RWMutex
```
Good idea! Currently for scaling purposes I think this has no downsides, so I changed to your suggestion with the latest push :)

ViktorTigerstrom commented on gbn/timeout_manager.go:
```diff
@@ -0,0 +1,142 @@
+	"sync"
+	"time"
+)
+
```
Thanks! Checkout the latest commit structure and see if it matches what you've requested here :)

ellemouton replied:
looks good!

ViktorTigerstrom commented on gbn/gbn_conn.go:
```diff
@@ -109,6 +90,10 @@ type GoBackNConn struct {
+
+	// timeoutManager is used to manage all the timeouts used by the
+	// GoBackNConn.
+	timeoutManager *TimeoutManager
```
Thanks, will keep that in mind from now on!

ViktorTigerstrom commented on mailbox/client_conn.go:
```diff
@@ -156,7 +160,7 @@ func NewClientConn(ctx context.Context, sid [64]byte, serverHost string,
 	c := &ClientConn{
 		transport: transport,
 		gbnOptions: []gbn.Option{
-			gbn.WithTimeout(gbnTimeout),
```
Yes this was ment to be a replace. If we use `WithTimeout`, the timeout will be set to a static value, and won't change dynamically even if we set a multiplier. So it's setting it as a static value, or letting it be dynamic :)

ellemouton replied:
gotcha! (btw - for comments like this where it's a question & not like a small nit or something - rather let the reviewer click "resolve comment" so that the reviewer doesnt miss the response to their question :)

ellemouton replied:
shall we then rather rename `WithTimeout` to `WithStaticTimeout` to make this clear?

ViktorTigerstrom replied:
Ok true, I've updated this with the latest push. Let me know if it addresses your feedback :)

ViktorTigerstrom replied:
> gotcha! (btw - for comments like this where it's a question & not like a small nit or something - rather let the reviewer click "resolve comment" so that the reviewer doesnt miss the response to their question :)

Ok good to know, thanks! Will keep that in mind from now on :)!

ViktorTigerstrom commented:
> Nice addition! Gonna be a game changer!

Thanks :fire::rocket:!! I've addressed your feedback with the latest push 🚀!


> * does this need to be based on [gbn: Sender side only packeting resend bug fix #87]
> (https://github.com/lightninglabs/lightning-node-connect/pull/87)? afaikt it doesnt depend on that change - so perhaps better to decouple them since it isnt clear which one will get in first yet.

I'll provide a more detailed answer to this tomorrow, but will provide a short answer for now. I should probably have mentioned this in the PR description previously, but the reason I've based this on #87, was that I noticed some potential compatibility issues when manually testing this, if only 1 of the 2 counterparties updated to a version of LNC that set the resend timeout to a different value than the default. It led to a resend loop of the packets that seemed potentially even worse than original packeting resend bug, but the fix in #87 largely fixed that if that party had also implemented that fix. I'll do more thorough testing tomorrow though, and see if my initial tests were correct.
 
> * A test for TimeoutMgr should please be added :)

This has been added now 🔥

lightninglabs-deploy commented:
@viktortigerstrom, remember to re-request review from reviewers when ready

ViktorTigerstrom commented on gbn/timeout_manager.go:
```diff
@@ -62,11 +73,76 @@ func NewTimeOutManager(isServer bool) *TimeoutManager {
+	// TODO(viktor): In the future, we may want to use this to keep track of
+	// the time it took for the other party to respond to other types of
+	// messages than the handshake, and dynamically keep updating the resend
+	// timeout to ensure that it reflects the current response time.
```
I've updated this PR to now also update the resend timeout on ACKs for `PackageData` messages which haven't been resent!

ViktorTigerstrom commented:
Added two new commits as the last commits of the PR that adds the following functionality:

When we need to resend a data packet, or when we need to resend the SYN message during the handshake, and the other side does not respond within given timeout, we will boost the timeouts by 50% for each time we need to resend without receiving any response.
This ensures that if the original timeouts are set to a too short duration given the current network latency, we will eventually boost the timeouts to a long enough duration that allows the other side to be able to respond within the timeout.

Note that this is currently also applied if to the resend timeout if we resend due to receiving a NACK from the other side due to packet loss during the transfer. If you'd like, I can refactor and pass along a bool that indicates if the resend happened due to a timeout or due to packet loss. I opted to not do that with the current implementation due to that there's no real tangible downside to temporarily boosting the resend timeout in such situations, and passing along the bool made things less clean.

guggero commented on gbn/gbn_conn.go:
```diff
@@ -46,9 +41,7 @@ type GoBackNConn struct {
-	sendTimeout time.Duration
-	recvTimeout time.Duration
-	timeoutsMu  sync.RWMutex
+	timeoutsMu sync.RWMutex
```
Do we still need this? Shouldn't all timeouts now be guarded by the timeout manager?

ViktorTigerstrom replied:
Good catch!

guggero commented on gbn/config.go:
```diff
@@ -2,6 +2,67 @@ package gbn
 
 import "time"
 
+type TimeoutOptions func(manager *TimeoutManager)
```
nit: missing Godoc.

guggero commented on gbn/timeout_manager.go:
```diff
@@ -17,6 +17,105 @@ const (
 	DefaultRecvTimeout            = math.MaxInt64
 )
 
+type TimeoutBooster struct {
```
nit: missing Godoc.

guggero commented on gbn/timeout_manager.go:
```diff
@@ -17,6 +17,105 @@ const (
+
+	increase := time.Duration(
+		float32(b.originalTimeout) * b.boostPercent *
+			float32(b.boostCount))
```
nit: closing parenthesis on new line.

guggero commented on gbn/config.go:
```diff
@@ -63,6 +63,15 @@ func WithKeepalivePing(ping, pong time.Duration) TimeoutOptions {
 	}
 }
 
+func WithBoostPercent(boostPercent float32) TimeoutOptions {
```
nit: missing Godoc.

ViktorTigerstrom commented on gbn/timeout_manager.go:
```diff
@@ -75,53 +79,121 @@ type TimeoutManager struct {
+			// the corresponding response, as we can't know if
+			// the response is for the resent SYN or the original
+			// SYN.
+			m.latestSentSYNTime = time.Unix(0, 0)
```
Ah, thanks :)!

ViktorTigerstrom commented on gbn/timeout_manager.go:
```diff
@@ -75,53 +79,121 @@ type TimeoutManager struct {
+	case *PacketACK:
+		m.sentTimesMu.Lock()
+		if sentTime, ok := m.sentTimes[msg.Seq]; ok {
+			m.sentTimesMu.Unlock()
```
True, good catch :)!

ViktorTigerstrom replied:
So this was specifically changed back to not be a `defer` again, to ensure that the lock is not held when calling the `updateResendTimeout` function. This is to ensure that we don't hold locks unnecessarily long that might introduce deadlocks in the future, to address another feedback by Oli :)

lightninglabs-deploy commented:
@guggero: review reminder
@ellemouton: review reminder

ellemouton commented on gbn/timeout_manager.go:
```diff
@@ -7,13 +7,14 @@ import (
-	defaultResendMultiplier   = 5
-	DefaultTMSendTimeout      = math.MaxInt64
-	DefaultTMRecvTimeout      = math.MaxInt64
+	defaultTMHandshakeTimeout     = 100 * time.Millisecond
```
i think this whole commit can be squashed with the previous one

ViktorTigerstrom replied:
Ok will keep that in mind for the future so I know where to draw the line of where to create a separate commit for something or not :).

ellemouton commented on gbn/timeout_manager_test.go:
```diff
@@ -46,14 +49,83 @@ func TestDynamicTimeout(t *testing.T) {
+	require.Equal(t, newTimeout, unchangedResendTimeout)
+}
+
+func TestDataPackageDynamicTimeout(t *testing.T) {
```
godoc?

ellemouton commented on gbn/gbn_conn.go:
```diff
@@ -251,21 +241,21 @@ func (g *GoBackNConn) start() {
 
 	pongTime := time.Duration(math.MaxInt64)
-	if g.cfg.pongTime != 0 {
-		pongTime = g.cfg.pongTime
+	if g.timeoutManager.GetPongTime() != 0 {
+		pongTime = g.timeoutManager.GetPongTime()
 	}
```
move all this logic into the timeout manager's GetPongTime msg?

ellemouton replied:
same with pingTime above

ViktorTigerstrom replied:
True thanks, it's been addressed by the latest push.

ellemouton commented on gbn/gbn_conn.go:
```diff
@@ -89,63 +84,62 @@ func newGoBackNConn(ctx context.Context, cfg *config,
 	prefix := fmt.Sprintf("(%s)", loggerPrefix)
 	plog := build.NewPrefixLog(prefix, log)
 
+	timeoutManager := NewTimeOutManager(true)
```
server = true for both?

ViktorTigerstrom replied:
Addressed to ensure the intermediate commits also has correct `is_server` values!

ellemouton commented on gbn/timeout_manager.go:
```diff
@@ -0,0 +1,277 @@
+	defaultTMFinSendTimeout   = 1000 * time.Millisecond
+	defaultResendMultiplier   = 5
+	DefaultTMSendTimeout      = math.MaxInt64
+	DefaultTMRecvTimeout      = math.MaxInt64
```
perhaps just move the existing one's here? that way we dont need to rename them again later on.

ellemouton commented on gbn/gbn_conn.go:
```diff
@@ -70,6 +61,10 @@ type GoBackNConn struct {
 	// remoteClosed is closed if the remote party initiated the FIN sequence.
 	remoteClosed chan struct{}
 
+	// timeoutManager is used to manage all the timeouts used by the
```
nice - I like this commit 👍

ellemouton commented on gbn/config.go:
```diff
@@ -2,6 +2,69 @@ package gbn
+// wait for ACKs before resending the queue.
+func WithResendTimeout(timeout time.Duration) TimeoutOptions {
+	return func(manager *TimeoutManager) {
+		manager.SetStaticResendTimeout(timeout)
```
can just set the value directly. Dont need to go through the "Set*" methods

ViktorTigerstrom replied:
See https://github.com/lightninglabs/lightning-node-connect/pull/88#discussion_r1429309650

ViktorTigerstrom replied:
Addressed with the latest push, thanks!

ellemouton commented on gbn/timeout_manager_test.go:
```diff
@@ -128,6 +141,121 @@ func TestDataPackageDynamicTimeout(t *testing.T) {
 	require.Equal(t, newResendTimeout, unchangedResendTimeout)
 }
 
+func TestResendBooster(t *testing.T) {
```
godoc

ellemouton reviewed (commented):
Leaving some comments for now - need to come back and look a bit deeper at the boosters. 

Looking good so far though! 💥

ViktorTigerstrom commented on gbn/timeout_manager.go:
```diff
@@ -0,0 +1,277 @@
+}
+
+// SetPongTime sets the pong time.
+func (m *TimeoutManager) SetPongTime(time time.Duration) {
```
Thanks, I've removed a bunch of them, but kept the ones that do something more than just setting the `TimeoutManager`'s internal values, and those that are indirectly exported by the gbn package. Let me know if this addresses your feedback :)!

ViktorTigerstrom commented on gbn/timeout_manager.go:
```diff
@@ -0,0 +1,277 @@
+
+// Received is should be called when a message is received by the connection.
+func (m *TimeoutManager) Received(msg Message) {
+	// TODO(viktor): In the future, we may want to use this to keep track of
```
Squashed the 2 first commits, so this was indirectly addressed by that.

guggero commented on gbn/timeout_manager.go:
```diff
@@ -0,0 +1,327 @@
+			// the corresponding response, as we can't know if
+			// the response is for the resent SYN or the original
+			// SYN.
+			m.latestSentSYNTime = time.Time{}
```
Do we need to guard this by a mutex as well? Since it looks to me that `Sent()` and `Received()` could be called concurrently?

guggero replied:
Same for the `m.useStaticTimeout`. I guess we guard the _write_ access to that but not the read access, so it could still cause a concurrent access issue.

ViktorTigerstrom replied:
> Do we need to guard this by a mutex as well? Since it looks to me that `Sent()` and `Received()` could be called concurrently?

Ah yeah, great catch! I've addressed this in the latest push.

> Same for the `m.useStaticTimeout`. I guess we guard the _write_ access to that but not the read access, so it could still cause a concurrent access issue.

Since the `useStaticTimeout` is only set through the functional options that's passed with the `TimeoutManager`'s constructor, this shouldn't be an issue.

ViktorTigerstrom replied:
I've now removed the `SetStaticResendTimout` function!

guggero commented on gbn/timeout_manager.go:
```diff
@@ -0,0 +1,327 @@
+
+	case *PacketACK:
+		m.sentTimesMu.Lock()
+		defer m.sentTimesMu.Unlock()
```
Could we just do the `sentTime, ok := m.sentTimes[msg.Seq]` part here then unlock `m.sentTimesMu`? I'm always careful to not call into methods (like `m.updateResendTimeout`) that potentially locks things while holding a lock within the caller. That can very easily lead to deadlocks if not careful. So this pattern would be easier to reason about.

ViktorTigerstrom replied:
Ah, yeah good idea :)! I'be updated this with the latest push :)!

guggero commented on gbn/gbn_conn.go:
```diff
@@ -556,8 +542,9 @@ func (g *GoBackNConn) receivePacketsForever() error { // nolint:gocyclo
-					g.cfg.resendTimeout*2 //nolint:gomnd
+
+				timeout := g.timeoutManager.GetResendTimeout()
+				recentlySent := sinceSent < timeout*2 //nolint:gomnd
```
Hmm, should we just disable the `gomnd` linter? Since normally we're pretty good with not using magic numbers anyway.

ViktorTigerstrom replied:
Thanks! I agree, so I added that disable to #87!

guggero reviewed (commented):
Awesome commit structure, code looks really good!
Two issues with potential locking conditions, otherwise this is very close.

Same question here about the integration tests though: Do we want them to pass in this PR already or are there further changes needed in follow-up PRs? Do we know what's causing the failures?

lightninglabs-deploy commented:
@ellemouton: review reminder
@viktortigerstrom, remember to re-request review from reviewers when ready

ViktorTigerstrom commented:
Thanks a lot for the review @guggero! I've addressed your feedback with the latest push.

> Same question here about the integration tests though: Do we want them to pass in this PR already or are there further changes needed in follow-up PRs? Do we know what's causing the failures?

I will fully investigate this before lunch tomorrow, but I'm quite confident that the issue why this has become worse since the addition of #87 (even though the issue existed prior to that), is that we will now end up waiting for the syncer to timeout in a bunch of the reconnection `itests`.

ViktorTigerstrom commented:
Rebased on the new version of #87.

ViktorTigerstrom commented:
As explained in [#87](https://github.com/lightninglabs/lightning-node-connect/pull/87#issuecomment-1876241391), the latest CI workflow now failed on the pre-existing issue (2).

ViktorTigerstrom commented:
Rebased on #87 which has now added the `itest` fix :rocket:

ViktorTigerstrom commented:
Rebased on new version of #87

ViktorTigerstrom commented:
Just rebased on master now that #87 has been merged, to make it clear which commits are part of this PR.

ViktorTigerstrom commented:
Hmm got a new `itest` error I haven't seen before. Seems likely to be caused by something in this PR. Looking into what's happening.

ViktorTigerstrom commented:
Ok so there turned out to be a few more issues effecting the itests, some that which were pre-existing and some some which where made a bit worse by this PR that now increases the resend timeout:

(1)
Now that we back-off and wait for the sync for potentially longer duration than the Ping ticker, If we have just sent a previous ping, both the pingTicker and pongTicker may have ticked when waiting to sync. That may cause that we continuously resend a ping without ever timing out due to the PongTicker Tick. I addressed this with: [fa371d8](https://github.com/lightninglabs/lightning-node-connect/pull/88/commits/fa371d8672c1c4aca30f06c192135763c2ebfd0a)

(2)
This issue the issue that happened earlier above, and was pre-existing, and I'd like to ensure that you agree that my solution to it is the correct way to handle the issue:
We had a bug that the server could end up in a scenario where it would not complete the handshake if the server restarted the handshake after sending its SYN, but before received the SYNACK response back from the client, before timing out due to latency or packet loss. If the client then received the SYN, it will proceeded to send the SYNACK and complete the handshake regardless of if the server receives the SYNACK in time or not.
If this happened, the server would then get stuck in the handshake while the client would leave the handshake phase.

I therefore changed the handshake logic for the server, so that if the server receives a SYNACK or PacketData after having sent it's SYN but timing out in the handshake, it would treat those as the completion of handshake. This was updated with [b447950](https://github.com/lightninglabs/lightning-node-connect/pull/88/commits/b447950e347939c16558f3a4f459709eb3ff00e1). Do you agree that this is a correct handling in the handshake? 

(3)
In itests, due to the client closing it's context before closing the connection, any FIN message sent by the client would always fail. I didn't really see any `itest` failures due to this, but I corrected it to ensure it doesn't cause any errors in the future with [6b5ca8b](https://github.com/lightninglabs/lightning-node-connect/pull/88/commits/6b5ca8b4895cb69efcca5ff13f6b5d378c725076).

@ellemouton, given the discussion in https://github.com/lightninglabs/lightning-node-connect/pull/87#discussion_r1442604224, is this way of ensuring that `c.cancel()` gets executed even if `c.grpcConn.Close()` errors ok as it closely resembles how we normally do with `returnErr` in itests? Or should I explicitly check the `c.grpcConn.Close()` and run `c.cancel()` if & if not an error was thrown.

(4)
Finally I bumped the `itest` timeout to 60seconds just a precaution for users which have a really bad connection 
256546e. This was done as I've realised that if a user get's the `resendTimeout` with boosting set to more than 10secs, the user is bound to timeout on the reconnection tests if unlucky. This is because if the other party (server or client) is down during the, and the syncer has just resent the packets, it will will then take 30 seconds (10 seconds X 3) for the syncing to time out and complete.
I should probably add that I haven't really had that bad latency so it hasn't been an issue for me yet, but I've been close to it when testing. Therefore i changed this as a precaution, but will drop this commit if you think it's an unnecessary addition.

guggero reviewed (approved):
Great investigations and bug hunting! Very excited to see the itests being stable with this PR as well.
Just one comment about a mutex, otherwise this LGTM :tada:

ellemouton commented on gbn/timeout_manager.go:
```diff
@@ -435,6 +493,8 @@ func (m *TimeoutManager) SetStaticResendTimeout(resendTimeout time.Duration) {
 	defer m.timeoutManagerMu.Unlock()
 
 	m.resendTimeout = resendTimeout
+	m.resendBooster.SetOriginalTimeout(resendTimeout)
```
instead set this timeout at the time of initialising the resendBooster. You can do this by just re-shuffling the  order of things in `NewTimeOutManager`: 

1: construct TimeoutMgr
2. Apply it's functional options.
3. initialise boosters

ViktorTigerstrom replied:
Thanks! Good idea, I've implemented this with the latest push. It required me to now add the `resendBoostPercent` & `handshakeTimeout` fields to the `TimeoutManager` though, which isn't ideal. But I think the new version is much better than the last version though, so I definitely think that's worth it.

ellemouton commented on gbn/gbn_conn.go:
```diff
@@ -413,11 +413,26 @@ func (g *GoBackNConn) sendPacketsForever() error {
 			continue
 
 		case <-g.pingTicker.Ticks():
+			// If we have awaited a sync after sending the previous
```
s/awaited/expected

ellemouton commented on gbn/gbn_server.go:
```diff
@@ -111,6 +112,24 @@ func (g *GoBackNConn) serverHandshake() error { // nolint:gocyclo
+				g.timeoutManager.Received(msg)
+
+				break handshakeLoop
+			} else {
```
no need for `else` block

ellemouton commented on gbn/gbn_server.go:
```diff
@@ -111,6 +112,24 @@ func (g *GoBackNConn) serverHandshake() error { // nolint:gocyclo
 		case *PacketSYN:
+
+		case *PacketSYNACK, *PacketData:
+			// If we receive a SYNACK or DATA packet after we have
+			// restarted the handshake, we can be sure that the
+			// client has received our SYN and has completed the
+			// handshake. We can therefore complete the handshake
+			// ourselves.
```
> we can be sure that the client has received our SYN

Hmmm no I dont think that is true.

What if the client is still on a previous connection resending Data with an Old `N` value? ie, it is not on a handshake at all? 

With the handshake, we need to not make any assumptions - we should be very sure that they are on the same page with the same N value. 

A couple of handshake retries is not a bad thing I think. 

let me know if I am missing something 🙏

ViktorTigerstrom replied:
Hmm I don't think that could become an issue with these changes?
Since the handshake process is:
1. Client: SYN
2. Server: SYN
3. Client: SYNACK

Since this newly added `if` case that allows the server to consider the client to have completed the handshake, only happens when the server has already received the client's SYN, we can be sure that the client isn't resending Data with an old `N`, as that can't happen after the client has sent the SYN. Or am I missing something?

What happened before is that the client could move on from the handshake, while the server would be stuck in it.
Just for context, I'm attaching the `itest` CI workflow that failed due to that:
https://github.com/lightninglabs/lightning-node-connect/actions/runs/7446471814/job/20256804763

Let me know though my understanding is incorrect, and this indeed could be an issue!

ellemouton replied:
ok cool 👍

ViktorTigerstrom commented on gbn/timeout_manager.go:
```diff
@@ -19,6 +19,109 @@ const (
+	defer b.boosterMu.Unlock()
+
+	b.originalTimeout = timeout
+}
```
Thanks! I've now removed the `SetOriginalTimeout` function completely, but didn't add a functional option as it's no longer needed due to the refactor suggested in https://github.com/lightninglabs/lightning-node-connect/pull/88#discussion_r1450248803 & the change of the `Reset` function.

ViktorTigerstrom commented on gbn/config.go:
```diff
@@ -2,6 +2,69 @@ package gbn
+// WithResendMultiplier is used to set the resend multiplier. This is the
+// multiplier we use when dynamically setting the resend timeout, based on how
+// long it took for other party to respond.
+// Note that when setting the resend timeout manually with the WithTimeout
```
Thanks, my bad!

ViktorTigerstrom commented on gbn/config.go:
```diff
@@ -2,6 +2,69 @@ package gbn
 
+// TimeoutOptions can be used to modify the default timeout values used within
+// the TimeoutManager.
+type TimeoutOptions func(manager *TimeoutManager)
```
Good idea, changed this with the latest push! It required me to temporarily rename the `WithHandshakeTimeout` & `WithKeepalivePing` options though, but I hope that's ok!

ViktorTigerstrom commented on gbn/gbn_client.go:
```diff
@@ -128,9 +128,11 @@ handshake:
 			default:
```
Thanks! The commit messages & order have been updated

ViktorTigerstrom commented on gbn/timeout_manager_test.go:
```diff
@@ -0,0 +1,200 @@
+	// 10-11 seconds as correct.
+	resendTimeout = tm.GetResendTimeout()
+	require.GreaterOrEqual(t, resendTimeout, time.Second*10)
+	require.Less(t, resendTimeout, time.Second*11)
```
Thanks, I'll keep that in mind from now on :)!

ViktorTigerstrom commented on gbn/timeout_manager_test.go:
```diff
@@ -0,0 +1,200 @@
+
+	// The resend timeout should now have dynamically changed.
+	resendTimeout := tm.GetResendTimeout()
+	require.NotEqual(t, initialResendTimeout, resendTimeout)
```
Changed this to `require.Greater` as we test the multiplier functionality right below. Hope that satisfies this feedback :)

ViktorTigerstrom commented on gbn/timeout_manager.go:
```diff
@@ -0,0 +1,353 @@
+
+	m.hasSetDynamicTimeout = true
+
+	multipliedTimeout := time.Duration(m.resendMultiplier) * responseTime
```
Ah thanks! Changed this

ViktorTigerstrom commented on gbn/timeout_manager.go:
```diff
@@ -0,0 +1,353 @@
+			m.updateResendTimeout(responseTime)
+		} else {
+			m.latestSentSYNTimeMu.Unlock()
+		}
```
Thanks, I've updated this throughout the PR! Hope this new version addresses your feedback 😃

ViktorTigerstrom commented on gbn/timeout_manager.go:
```diff
@@ -0,0 +1,353 @@
+	defer m.timeoutManagerMu.Unlock()
+
+	m.handshakeTimeout = handshakeTimeout
+}
```
Those functions have now been removed!

ViktorTigerstrom commented:
Thanks a lot for the reviews @guggero & @ellemouton!
I've addressed your feedback with the latest push!

@ellemouton, please specifically check my response to https://github.com/lightninglabs/lightning-node-connect/pull/88#discussion_r1451870412, as that wasn't addressed with this latest push!

ellemouton commented on gbn/timeout_manager.go:
```diff
@@ -0,0 +1,338 @@
+
+	// timeoutManagerMu should be locked when updating or accessing any of
+	// timeout manager's timeout fields.
+	timeoutManagerMu sync.RWMutex
```
nit: stuttering: "TimeoutManager.timeoutManagerMu" -> just mu :)

ellemouton commented on gbn/timeout_manager.go:
```diff
@@ -0,0 +1,338 @@
+
+		m.sentTimesMu.Unlock()
+
+		m.responseCounter++
+
+		reachedFrequency := m.responseCounter%
+			m.timeoutUpdateFrequency == 0
```
isn't there a possible race on access to `responseCounter` here? nothing is stopping Receive from being called concurrently.

ellemouton replied:
should the main timeoutMgr's mutex not be held for the duration of Receive and not only when updating the resend timeout?

ViktorTigerstrom replied:
Thanks, good idea!

So currently, since the `GoBackNConn` `sendPacketsForever` & `receivePacketsForever` are individually run synchronously, but concurrent from each other, we will never get multiple concurrent `TimeoutManager` `Sent` & `Received` calls, they can only happen concurrent from each other. But I updated to protect the `responseCounter` by the `TimeoutManager` `mu` regardless, as in case the above would ever change in the future :)!

> should the main timeoutMgr's mutex not be held for the duration of Receive and not only when updating the resend timeout?

I actually don't really agree with this though. The reason for that is that in case the `updateResendTimeout` (which also accesses the same `mu`) ever get's more complicated & slow in the future, the measurement of the time it took between a `Sent` & `Received` message might get skewed in the future, as `Received` might then get stuck at accessing the `TimeoutManager` `mu` for an extended time.

ellemouton replied:
> are individually run synchronously, but concurr

Yeah we should not depend on the API being used in a synchronous way. 

do we really ever expect `updateResendTimeout` to take long though? 

> he measurement of the time it took between a Sent & Received message might get skewed in the future, as Received might then get stuck at accessing the TimeoutManager mu for an extended time.

I mean, you could just do the `time.Now()` before acquiring the lock so that the time you get the message is still accurate. You could also write a benchmark test to measure the performance of the current impl vs one with just a single mutex to see if it really does make a big difference if you wanted to verify this btw.  [EDIT: Ok I wrote one. see below]

Currently: the `mu` lock is held here and then immediately acquired again in `updateResendTimeout` which is a red flag for a race condition. 

Also just in general having all these locks locked/unlocked at different times without a single long running lock is a bit of a flag for a race condition. Like above we now have this: 

```
		m.sentTimesMu.Unlock()

		m.mu.Lock()
```

which can cause a race if `Received` were to be called by 2 threads at the same time. 

same with the above case: 
```
m.latestSentSYNTimeMu.Unlock()

m.updateResendTimeout(responseTime)
```

Just for fun, i wrote some basic benchmark tests to compare things. So yes, if we must `mu` everywhere for everything, things are a bit slower (around 80ns/op slower). So cool - let's keep the individual mutexes. But let's fix the race's. When I fixed the races the
 benchmark tests show no performance difference. See the patch here: [timeoutmgr.patch](https://github.com/lightninglabs/lightning-node-connect/files/13950552/timeoutmgr.patch)
I've left the benchmark tests there so you can see what they look like (but feel free to delete. Might be worth keeping the second one just for quick testing of any future modifications)

ViktorTigerstrom replied:
Thanks a lot for the great benchmarks and the patch @ellemouton! I've applied it and added it with slight modifications :)!

> do we really ever expect updateResendTimeout to take long though?

No not currently at least. What I tried to say is that it shouldn't effect the measurement of the time it took between a `Sent` & `Received` message even if it were to change in the future. Your latest patch addresses that potential issue though :)!

> you could just do the time.Now() before acquiring the lock

Good idea, thanks :tada:!

> Also just in general having all these locks locked/unlocked at different times without a single long running lock is a bit of a flag for a race condition. Like above we now have this:

Oh I actually thought it was the opposite, TIL then :)! I thought you should avoid holding locks while when acquiring other locks, unless necessary, as that creates lock orders, which then makes it very easy to introduce deadlocks in the future if your not careful.
In terms of the approach when releasing locks before acquiring new ones potentially leading to different threads acquiring the locks before the other threads that attempt to acquire the same lock, I thought that was what's expected in concurrent programming. It could also potentially lead to more optimized asynchronous execution if designed correctly. As in a case where you have a function `func1` that acquires `lock1` followed by `lock2`, if the `lock1` is held while acquiring `lock2` and then releases `lock1`, if a thread has executed `func1` and is temporarily stuck at waiting to acquire `lock2`, any concurrent thread that executes `func1` will now need to wait for the first thread to acquire `lock2` before the code that requires `lock1` can be executed, potentially slowing the overall execution.

I've changed these cases to what your suggestion, but in case my old understanding is valid, let me know and I'll change back :)!

> i wrote some basic benchmark tests to compare things

Awesome, thanks a lot :pray::rocket:!! I kept both of them, as I think it adds extra coverage and it doesn't hurt to have both a synchronous and a concurrent version. Let me know if you really think I should remove the first one though!

ellemouton replied:
Your understanding is 100% correct but it just means that we should make sure that the locks are not acquired an a different order else-where (which would lead to a deadlock) and that they are not held for very long. The other option is just to have a single mutex and use that for everything but as you've pointed out, that would slow things down a bit. So if we are going with multiple mutexes then that means we need to just be very careful _but_ it still needs to be correct behaviour. Releasing one lock before acquiring another can lead to undefined behaviour if the  API is used in a multithreaded way.

ellemouton replied:
>  I kept both of them, as I think it adds extra coverage and it d
Sounds good!

ellemouton commented on itest/client_harness.go:
```diff
@@ -119,6 +119,9 @@ func (c *clientHarness) start() error {
 
 func (c *clientHarness) cleanup() error {
-	c.cancel()
+	// we always want to cancel the context, even if
```
s/we/We

Also - this change is more about canceling the context after closing the conn. Not about closing it regardless of error.

ViktorTigerstrom replied:
Thanks, updated this message :)!

ellemouton reviewed (approved):
LGTM pending one last comment regarding a possible race. 

Also tested this with old server & new client & a restarting mailbox 👍 

Really great work 🎉

ViktorTigerstrom commented:
Thanks a lot for the review and testing @ellemouton :rocket::fire:!!

Let me know if you disagree with https://github.com/lightninglabs/lightning-node-connect/pull/88#discussion_r1453334124, else this should be good to go :tada::rocket:!!

ellemouton commented on gbn/timeout_manager.go:
```diff
@@ -0,0 +1,522 @@
+
+	m.resendTimeout = multipliedTimeout
+
+	// Also update and reset the resend booster, as the new dynamic
+	// resend timeout most accurately reflects the current response
+	// time.
+	m.resendBooster.Reset(multipliedTimeout)
+
+	// As we may have received a data packet that executes this function
+	// while we are also concurrently resending the queue, we also restart
+	// the frequency timeout, to ensure that the messages we're resending
+	// won't boost the resend timeout.
+	m.resendBooster.RestartFrequencyTimeout()
```
one last thing:

these two are both only ever called together & are only ever called here. But each of them acquire the booster's mutex individually. So why not instead compile the two? `ResetAndRestart` or something like that? then that lock only needs to be acquired once

ViktorTigerstrom replied:
Sure! I opted to just include the `RestartFrequencyTimeout` in the `Reset` function, but instead keep `Reset` name with an added comment. Hope you agree about that approach :)!

ellemouton commented on gbn/timeout_manager_test.go:
```diff
@@ -0,0 +1,369 @@
+
+// BenchmarkTimeoutMgrConcurrently benchmarks the timeout manager when sending
+// and receiving messages concurrently.
+func BenchmarkTimeoutMgrConcurrently(b *testing.B) {
```
nice! 

ok last last thing I promise 😂  can we add a TestStressTestTimoutMgr test that basically does this exact test with  very high N value. This will just make sure that this stress test is part of our CI and then if we ever actually do have a deadlock situation I dont think bench mark tests are currently run by the CI but I could be wrong), it would likely get caught by the test. it can literally just be:

```
func TestStressTestTimeoutMgr(t *testing.T) {
	tm := NewTimeOutManager(nil, WithTimeoutUpdateFrequency(2))

	var wg sync.WaitGroup
	for n := 0; n < 100000; n++ {
		wg.Add(1)
		go func(seq uint8) {
			defer wg.Done()

			msg := &PacketData{Seq: seq}

			tm.Sent(msg, false)
			tm.Received(msg)
		}(uint8(n))
	}

	wg.Wait()
}
```

ViktorTigerstrom replied:
Ah good idea! Added the test with a slight modification to ensure it runs in parallel with the other tests :)!

ellemouton reviewed (approved):
Awesome stuff @ViktorTigerstrom!!!

Ok this is basically good to go. Added 2 final thoughts :)

ViktorTigerstrom commented:
After discussing this with @ellemouton offline, I've updated so that the `TimeoutManager's` `Received` function grabs the main `mu` in the beginning of the function. This ensures that any `GetResendTimeout` call we receive concurrently after the
one thread has just gotten a `Received` call that will update the resend timeout, will then return the updated timeout!

ellemouton reviewed (approved):
Awesome work!
